SofaKernel/modules/SofaDeformable/src/SofaDeformable/SpringForceField.inl

```diff
@@ -89,13 +89,119 @@ void SpringForceField<DataTypes>::reinit()
+        modifiedTopology = this->getMState1()->getContext()->getMeshTopology();
+    }
+    else
+    {
+        modifiedTopology = this->getMState2()->getContext()->getMeshTopology();
+    }
+
+    if (modifiedTopology == nullptr)
+        return;
+
+    const auto& tab = pointsRemoved->getArray();
+
+    if (tab.empty())
+        return;
+
+    type::vector<Spring>& springsValue = *sofa::helper::getWriteAccessor(this->springs);
+    auto nbPoints = modifiedTopology->getNbPoints();
+
+    for (const auto pntId : tab)
+    {
+        --nbPoints;
+
+        sofa::type::vector<sofa::Index> toDelete;
+        sofa::Index i {};
+        for (const auto& spring : springsValue)
+        {
+            auto& id = mstateId == 0 ? spring.m1 : spring.m2;
+            if (id == pntId)
+            {
+                dmsg_info() << "Spring " << spring << " has a removed point: REMOVED";
+                toDelete.push_back(i);
+            }
+            ++i;
+        }
+
+        for (auto it = toDelete.rbegin(); it != toDelete.rend(); ++it)
+        {
+            springsValue.erase(springsValue.begin() + (*it));
+        }
+
+        for (auto& spring : springsValue)
+        {
+            auto& id = mstateId == 0 ? spring.m1 : spring.m2;
+            if (id == nbPoints)
+            {
+                dmsg_info() << "Spring " << spring << " has a renumbered point: MODIFY from " << id << " to " << pntId;
+                id = pntId;
+            }
+        }
+    }
+}
+
 template <class DataTypes>
 void SpringForceField<DataTypes>::init()
 {
     // Load
     if (!fileSprings.getValue().empty())
         load(fileSprings.getFullPath().c_str());
     this->Inherit::init();
+
+    initializeTopologyHandler(d_springsIndices[0], this->mstate1->getContext()->getMeshTopology(), 0);
+    initializeTopologyHandler(d_springsIndices[1], this->mstate2->getContext()->getMeshTopology(), 1);
+
+    updateTopologyIndicesFromSprings();
+}
+
+template <class DataTypes>
+void SpringForceField<DataTypes>::initializeTopologyHandler(sofa::core::topology::TopologySubsetIndices& indices,
+    core::topology::BaseMeshTopology* topology, sofa::Index mstateId)
+{
+    if (topology)
+    {
+        indices.createTopologyHandler(topology);
+
+        indices.addTopologyEventCallBack(core::topology::TopologyChangeType::POINTSREMOVED,
+            [this, mstateId](const core::topology::TopologyChange* change)
+            {
+                const auto* pointsRemoved = static_cast<const core::topology::PointsRemoved*>(change);
+                msg_info(this) << "Removed points: [" << pointsRemoved->getArray() << "]";
+                applyRemovedPoints(pointsRemoved, mstateId);
+            });
+        indices.addTopologyEventCallBack(core::topology::TopologyChangeType::ENDING_EVENT,
+            [this](const core::topology::TopologyChange*)
+            {
+                msg_info(this) << "Update topology indices from springs";
+                updateTopologyIndicesFromSprings();
```

Comment:
I don't understand why tracking POINTSREMOVED event and updating the springs using applyRemovedPoints if after ENDING_EVENT you recompute every springs ?
You should either try to track in a smart way using POINTSREMOVED  / POINTSADDED events
or easier but slower, recompute every springs each time a ENDING_EVENT  is called.

Reply:
There are 2 different structures (Data):
1. `Data<sofa::type::vector<Spring> > springs;`
2. `std::array<sofa::core::topology::TopologySubsetIndices, 2> d_springsIndices`.

`d_springsIndices` is just a "copy" of the indices in `springs`. `d_springsIndices` exists just to track topological changes. Otherwise, I don't know how to add the tracking on `springs`. Therefore, I added a first pass POINTSREMOVED to update `springs` from the changes in `d_springsIndices`. And a second pass ENDING_EVENT to update `d_springsIndices` from the springs. The second pass is required to recompute both elements of `d_springsIndices`. Otherwise, only one of the 2 is modified. Does it make sense?

Reply:
@epernod I need this PR in https://github.com/sofa-framework/sofa/pull/2554. I would appreciate your feedback on this before merging it.